### PR TITLE
KF-3793 Add istio relation

### PR DESCRIPTION
**Pull request by misohu**
* Add istio relation

### CODE REVIEW

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -114,6 +117,51 @@ def namespace(lightkube_client: lightkube.Client):
     delete_all_from_yaml(yaml_text, lightkube_client)
 
 
+async def setup_istio(ops_test: OpsTest, istio_gateway: str, istio_pilot: str):
+    """Deploy Istio Ingress Gateway and Istio Pilot."""
+    await ops_test.model.deploy(
+        entity_url="istio-gateway",
+        application_name=istio_gateway,
+        channel="latest/edge",
```
Comment (beliaev-maksim) [inline tests/integration/test_charm.py:125]: are we sure with edge?

File: src/charm.py [hunk 2]
```diff
@@ -366,6 +366,17 @@ def _create_manifests(self, manifest_files, context):
...
         return json.dumps(manifests)
 
+    def _send_ingress_info(self, interfaces):
+        if interfaces["ingress"]:
+            interfaces["ingress"].send_data(
+                {
+                    "prefix": "/mlflow/",
+                    "rewrite": "/",
+                    "service": self.model.app.name,
+                    "port": int(self._port),
```
Comment (ca-scribner) [inline src/charm.py:376]: Let's add the namespace here too.  If we omit it, then it assumes you're in the same namespace as istio-pilot which might not be true.  
```suggestion
                    "prefix": "/mlflow/",
                    "rewrite": "/",
                    "service": self.model.app.name,
                    "namespace": self.model.name,
                    "port": int(self._port),
```

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -114,6 +117,51 @@ def namespace(lightkube_client: lightkube.Client):
...
+        apps=[istio_pilot, istio_gateway],
+        status="active",
+        timeout=60 * 5,
+    )
+
+
+@pytest.fixture
+@pytest.mark.asyncio
+async def url_with_ingress(ops_test: OpsTest):
+    command = "microk8s kubectl get svc -n testing --output=jsonpath='{.status.loadBalancer.ingress[0].ip}' istio-ingressgateway-workload"  # noqa: E501
```
[the remark this answers is not in the document]
Comment (ca-scribner) [reply tests/integration/test_charm.py:147]: oh here's a simpler one.  steal [this](https://github.com/canonical/bundle-kubeflow/blob/517d6475d812d2f5d7b48b93ed4663d4d8942fb9/tests-bundle/1.6/helpers.py#L4)

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -362,3 +410,14 @@ async def test_can_connect_exporter_and_get_metrics(self, ops_test: OpsTest):
         assert 'mlflow_metric{metric_name="num_experiments"} 2.0' in metrics_text
         assert 'mlflow_metric{metric_name="num_registered_models"} 0.0' in metrics_text
         assert 'mlflow_metric{metric_name="num_runs"} 0' in metrics_text
+
+    @pytest.mark.abort_on_fail
+    async def test_ingress_relation(self, ops_test: OpsTest):
+        """Setup Istio and relate it to the Tensoboards Web App(TWA)."""
```
Comment (ca-scribner) [inline tests/integration/test_charm.py:416]:
```suggestion
        """Setup Istio and relate it to the MLflow."""
```

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -362,3 +410,14 @@ async def test_can_connect_exporter_and_get_metrics(self, ops_test: OpsTest):
...
+        await setup_istio(ops_test, ISTIO_GATEWAY_CHARM_NAME, ISTIO_PILOT_CHARM_NAME)
+
+        await ops_test.model.add_relation(
+            f"{ISTIO_PILOT_CHARM_NAME}:ingress", f"{CHARM_NAME}:ingress"
+        )
+
+        await ops_test.model.wait_for_idle(apps=[CHARM_NAME], status="active", timeout=60 * 5)
```
Comment (ca-scribner) [inline tests/integration/test_charm.py:423]: can we add a check to see if we can actually get to the mlflow page through the ingress?

Comment (ca-scribner) [reply tests/integration/test_charm.py:423]: I think you meant to include this check but it got missed?  I see fetch_response and url_with_ingress are unused

**Review by ca-scribner — CHANGES_REQUESTED** [review]
Almost there.  A few changes requested

**Review by ca-scribner — APPROVED** [review]
love it! lgtm

**Pull request merged**